### ISSUE TITLE
sympy.solve returns imaginary solutions contrary to given assumptions

### ISSUE THREAD

**Issue by woodscn** [sympy/sympy#9973]
```
>>> import sympy
>>> x = sympy.Symbol('x', real=True)
>>> y = sympy.Symbol('y', real=True)
>>> z = sympy.Symbol('z', real=True)
>>> eq = sympy.Eq(x**2 + y**2 + z**2 - 1, 0)
>>> sympy.solve(eq.subs({y: 1}), x)
[-I*z, I*z]
```
Given the assumptions for x, this result is clearly wrong. For this problem, I would expect something more like this:  
```
x == z == 0
```

**Comment by woodscn** [comment]
I should mention I'm interested in doing work on this, if someone can point me in the right direction.

**Comment by asmeurer** [comment]
What does it return if no assumptions are set?

**Comment by woodscn** [comment]
Same result.

**Comment by smichr** [comment]
The solver tries to determine solutions for x that don't violate assumptions. The two results are correct in that regard. The solver doesn't try to figure out what values of z will make the the solution match the assumptions on x, however, and I'm not sure that one should change "check_assumptions" to do so since solving for values satisfying the assumptions is beyond to scope of checking that the assumptions are not violated. But perhaps when there are assumptions on x (as in your case) another function which tries to reduce a set of expressions to the values that would make them true might be in order, e.g. given expressions (e1..en) univariate in x what values of x make them all real...or positive...or something else. For example, for a single expression:

```
>>> var('x', real=True)
x
>>> e=I*x
>>> Union(solve(e>0).as_set(),solve(e<0).as_set(),Eq(e,0).as_set()) # e is real
{0}
>>> Union(solve(e>0).as_set()) # e is positive
EmptySet()
>>> Union(solve(e>0).as_set(),Eq(e,0).as_set()) # e is nonnegative
{0}
```

**Comment by woodscn** [comment]
I don't think I agree that this is out of scope. I am asking solve to take an expression, and to rearrange it based on the assumptions inherent in the variables. There's no difference in the expression I'm handing sympy between variables X, Y, and Z, so why should solve ignore the assumptions on one of them but not the other? 

To put another way, solve was given a well-posed problem with a unique, specific, solution, but it stopped halfway through solving it. Instead, solve should do everything possible to return the simplest possible answer. In this case, that is simple, since the product of an imaginary number and a non-zero real number will always be imaginary, and solve knows that is not a solution to the problem it was given. 

Individuals can certainly enforce this themselves, using something like the example you showed. Unfortunately, this is not really a good solution if you are using sympy for code generation or anything else that needs to be automatic.

**Comment by smichr** [comment]
I agree. I am just saying that it should probably be a function of its own, not an add-on to `check_assumptions`.
